nifi-nar-bundles/nifi-standard-services/nifi-dbcp-service-bundle/nifi-dbcp-service/src/main/java/org/apache/nifi/dbcp/DBCPConnectionPool.java

```diff
@@ -413,6 +437,15 @@ protected ClassLoader getDriverClassLoader(String locationString, String drvName
      */
     @OnDisabled
     public void shutdown() {
+        try {
```

Comment:
Do we want to still try `dataSource.close()` even when `kerberos.logout()` throws an exception? or maybe we want to reverse these so that it would be the reverse of what we do in @OnEnabled ?

Right now it would bounce out of the method because of the ProcessException.

Reply:
Good point.  I moved the logout so that it happens after the datasource is closed.  However, I'm not certain that we want to throw exceptions if either of those operations fail.  If an exception is thrown, it will prevent the controller service from stopping, right?  I checked the HiveaConnectionPool's `@OnDisabled` method, which throws an exception if the datasource can't be closed, but some other controller services log the exception and allow the method to complete, which would allow the service to stop.

What are your thoughts on doing that with the DBCPConnectionPool?  If we allow the `@OnDisabled` method to complete even after an exception, the next time the controller services is started, the member variables should be overwritten and eventually garbage collected (or we could set them all to null before the method finishes), but I'm not sure we want to rely on that.

It seems worth thinking about, because we don't want to prevent the service from stopping without a way to stop it eventually, without restarting NiFi.

Reply:
I would probably recommend using a `try/finally` to ensure that we do both. It's okay to throw an Exception from a method with `@OnDisabled` - from the docs:
```
If a method with this annotation throws a Throwable, a log message and
bulletin will be issued for the service, but the service will still be marked as Disabled.
The failing method will not be called again until the service is enabled and disabled again.
This is done in order to prevent a ControllerService from continually failing
in such a way that the service could not be disabled and updated without
restarting the instance of NiFi.
```

Reply:
@markap14 Sounds like there's no reason to throw the exceptions that occur in an `@OnDisabled` method.  Instead, I can just log them as errors and let the method finish.

Reply:
Yup, absolutely. Either approach is perfectly acceptable, IMO

Reply:
Pushed a new commit that should take care of attempting to close the datasource and log out the principal, with the datasource exception being prioritized over the logout exception, if both occur.